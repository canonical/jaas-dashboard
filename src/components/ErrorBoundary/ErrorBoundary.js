import { Component } from "react";
import * as Sentry from "@sentry/browser";

export default class ErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error) {
    return {
      hasError: true,
      error,
    };
  }

  componentDidCatch(error, info) {
    Sentry.withScope((scope) => {
      scope.setExtras(info);
      const eventId = Sentry.captureException(error);
      this.setState({ eventId });
    });
  }

  render() {
    const { hasError } = this.state;
    const { children } = this.props;
    if (hasError) {
      return (
        <div className="p-notification--negative">
          <p className="p-notification__response">
            <span className="p-notification__status">Error: </span>Something has
            gone wrong. If this issue persists,{" "}
            <a href="https://github.com/canonical-web-and-design/jaas-dashboard/issues/new?assignees=&labels=&template=bug_report.md&title=">
              please raise an issue on GitHub.
            </a>
          </p>
        </div>
      );
    }

    return <>{children}</>;
  }
}
