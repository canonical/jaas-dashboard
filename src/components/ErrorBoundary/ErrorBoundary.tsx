import { Component, PropsWithChildren } from "react";
import * as Sentry from "@sentry/browser";
import {
  Notification,
  Strip,
  CodeSnippet,
  CodeSnippetBlockAppearance,
} from "@canonical/react-components";
import { Extras } from "@sentry/types";

type Props = PropsWithChildren;

type State = {
  error?: Error;
  eventId?: string;
  hasError: boolean;
};

export default class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error: Error) {
    return {
      hasError: true,
      error,
    };
  }

  componentDidCatch(error: Error, info: unknown) {
    Sentry.withScope((scope) => {
      scope.setExtras(info as Extras);
      const eventId = Sentry.captureException(error);
      this.setState({ eventId });
    });
  }

  render() {
    const { error, hasError } = this.state;
    const { children } = this.props;
    const body = encodeURIComponent(
      `\`\`\`\n${error?.stack ?? "No stack track"}\n\`\`\``
    );
    const url = `https://github.com/canonical-web-and-design/jaas-dashboard/issues/new?assignees=&labels=&template=bug_report.md&title=Dashboard error: ${encodeURIComponent(
      error?.message ?? "No error"
    )}&body=${body}`;
    if (hasError) {
      return (
        <Strip>
          <Notification severity="negative" title="Error">
            Something has gone wrong. If this issue persists,{" "}
            <a href={url} rel="noopener noreferrer" target="_blank">
              please raise an issue on GitHub.
            </a>
          </Notification>
          <CodeSnippet
            blocks={[
              ...(error?.message
                ? [
                    {
                      title: "Error",
                      appearance: CodeSnippetBlockAppearance.NUMBERED,
                      wrapLines: true,
                      code: error.message,
                    },
                  ]
                : []),
              ...(error?.stack
                ? [
                    {
                      title: "Stack trace",
                      appearance: CodeSnippetBlockAppearance.NUMBERED,
                      wrapLines: true,
                      code: error.stack,
                    },
                  ]
                : []),
            ]}
          />
        </Strip>
      );
    }

    return <>{children}</>;
  }
}
