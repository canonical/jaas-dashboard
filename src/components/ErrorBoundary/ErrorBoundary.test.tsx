import { render } from "@testing-library/react";
import type { MockInstance } from "vitest";
import { vi } from "vitest";

import ErrorBoundary from "./ErrorBoundary";

function ChildComponent() {
  const error = new Error("Oh noes!");
  error.stack = "Stack trace output";
  throw error;
  // eslint-disable-next-line no-unreachable
  return null;
}

describe("Error Boundary", () => {
  let consoleError: MockInstance;

  beforeEach(() => {
    // Vitest does not catch the error thrown in the component so this cleans up
    // the console output.
    consoleError = vi.spyOn(console, "error").mockImplementation(() => vi.fn());
  });

  afterEach(() => {
    consoleError.mockRestore();
  });

  it("renders without crashing and matches snapshot", () => {
    const { container } = render(
      <ErrorBoundary>
        <ChildComponent />
      </ErrorBoundary>,
    );
    expect(container).toMatchSnapshot();
  });
});
