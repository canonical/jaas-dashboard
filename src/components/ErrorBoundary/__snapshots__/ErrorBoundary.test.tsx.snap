// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Boundary renders without crashing and matches snapshot 1`] = `
<div>
  <div
    class="p-notification--negative"
  >
    <p
      class="p-notification__response"
    >
      <span
        class="p-notification__status"
      >
        Error: 
      </span>
      Something has gone wrong. If this issue persists,
       
      <a
        href="https://github.com/canonical-web-and-design/jaas-dashboard/issues/new?assignees=&labels=&template=bug_report.md&title="
      >
        please raise an issue on GitHub.
      </a>
    </p>
  </div>
</div>
`;
